File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/ProgressBar.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;

import java.awt.*;
import java.util.List;

/**
 * LoadingScreen is designed to show progress the execution of any long time
 * task.
 * <p>
 * Contains image and text.
 * <p>
 * Supports all events except drag and drop.
 * <p>
 * By default ProgressBar cannot get focus.
 */
public class ProgressBar extends Prototype {
    private static int count = 0;
    private TextLine _textObject;

    /**
     * Setting the text that represents the progress of the unfinished task, visible
     * or invisible.
     * 
     * @param value True: if text should be visible. False: if text should be
     *              invisible.
     */
    public void setValueVisible(boolean value) {
        _textObject.setVisible(value);
    }

    /**
     * Returns True if text that represents the progress of the unfinished task is
     * visible, otherwise returns False.
     * 
     * @return True: if text is visible. False: if text is invisible.
     */
    public boolean isValueVisible() {
        return _textObject.isVisible();
    }

    private Rectangle _rect;
    private int _maxValue = 100;
    private int _minValue = 0;
    private int _currentValue = 0;

    /**
     * Default ProgressBar constructor.
     */
    public ProgressBar() {
        isFocusable = false;
        setItemName("ProgressBar_" + count);
        count++;

        _textObject = new TextLine();
        _textObject.setItemName(getItemName() + "_textObject");
        _textObject.setItemText("0%");
        _rect = new Rectangle();

        setStyle(DefaultsService.getDefaultStyle(ProgressBar.class));
    }

    /**
     * Initializing all elements in the ProgressBar.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        addItems(_rect, _textObject);
    }

    /**
     * Setting the maximum progress value of the unfinished task limit. Progress
     * value cannot be greater than this limit.
     * 
     * @param value Maximum progress value of the unfinished task limit.
     */
    public void setMaxValue(int value) {
        _maxValue = value;
    }

    /**
     * Getting the current maximum progress value of the unfinished task limit.
     * 
     * @return Maximum progress value of the unfinished task limit.
     */
    public int getMaxValue() {
        return _maxValue;
    }

    /**
     * Setting the minimum progress value of the unfinished task limit. Progress
     * value cannot be less than this limit.
     * 
     * @param value Minimum progress value of the unfinished task limit.
     */
    public void setMinValue(int value) {
        _minValue = value;
    }

    /**
     * Getting the current minimum sprogress value of the unfinished task limit.
     * 
     * @return Minimum progress value of the unfinished task limit.
     */
    public int getMinValue() {
        return _minValue;
    }

    /**
     * Setting the current progress value of the unfinished task. If the value is
     * greater/less than the maximum/minimum progress value, then the progress value
     * becomes equal to the maximum/minimum value.
     * 
     * @param currentValue Progress value of of the unfinished task.
     */
    public void setCurrentValue(int currentValue) {
        _currentValue = currentValue;
        updateProgressBar();
    }

    /**
     * Getting the progress value of the unfinished task.
     * 
     * @return Progress value of the unfinished task.
     */
    public int getCurrentValue() {
        return _currentValue;
    }

    private void updateProgressBar() {
        float AllLength = _maxValue - _minValue;
        float DonePercent;
        _currentValue = (_currentValue > _maxValue) ? _maxValue : _currentValue;
        _currentValue = (_currentValue < _minValue) ? _minValue : _currentValue;
        DonePercent = (_currentValue - _minValue) / AllLength;
        String text = Math.round(DonePercent * 100f) + "%";
        _textObject.setItemText(text);
        _rect.setWidth((int) Math.round(getWidth() * DonePercent));
    }

    /**
     * Setting alignment of a ProgressBar text. Combines with alignment by
     * vertically (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(ItemAlignment... alignment) {
        _textObject.setTextAlignment(alignment);
    }

    /**
     * Setting alignment of a ProgressBar text. Combines with alignment by
     * vertically (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as List of
     *                  com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(List<ItemAlignment> alignment) {
        _textObject.setTextAlignment(alignment);
    }

    /**
     * Setting indents for the text to offset text relative to ProgressBar.
     * 
     * @param margin Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public void setTextMargin(Indents margin) {
        _textObject.setMargin(margin);
    }

    /**
     * Setting indents for the text to offset text relative to ProgressBar.
     * 
     * @param left   Indent on the left.
     * @param top    Indent on the top.
     * @param right  Indent on the right.
     * @param bottom Indent on the bottom.
     */
    public void setTextMargin(int left, int top, int right, int bottom) {
        _textObject.setMargin(left, top, right, bottom);
    }

    /**
     * Getting indents of the text.
     * 
     * @return Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public Indents getTextMargin() {
        return _textObject.getMargin();
    }

    /**
     * Setting font of the text.
     * 
     * @param font Font as java.awt.Font.
     */
    public void setFont(Font font) {
        _textObject.setFont(font);
    }

    /**
     * Setting font size of the text.
     * 
     * @param size New size of the font.
     */
    public void setFontSize(int size) {
        _textObject.setFontSize(size);
    }

    /**
     * Setting font style of the text.
     * 
     * @param style New font style (from java.awt.Font package).
     */
    public void setFontStyle(int style) {
        _textObject.setFontStyle(style);
    }

    /**
     * Setting new font family of the text.
     * 
     * @param fontFamily New font family name.
     */
    public void setFontFamily(String fontFamily) {
        _textObject.setFontFamily(fontFamily);
    }

    /**
     * Getting the current font of the text.
     * 
     * @return Font as java.awt.Font.
     */
    public Font getFont() {
        return _textObject.getFont();
    }

    /**
     * Getting the text width (useful when you need resize ProgressBar by text
     * width).
     * 
     * @return Text width.
     */
    public int getTextWidth() {
        return _textObject.getWidth();
    }

    /**
     * Getting the text height (useful when you need resize ProgressBar by text
     * height).
     * 
     * @return Text height.
     */
    public int getTextHeight() {
        return _textObject.getHeight();
    }

    /**
     * Setting text color of a ProgressBar.
     * 
     * @param color Text color as java.awt.Color.
     */
    public void setForeground(Color color) {
        _textObject.setForeground(color);
    }

    /**
     * Setting text color of a ProgressBar in byte RGB format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b) {
        _textObject.setForeground(r, g, b);
    }

    /**
     * Setting text color of an item in byte RGBA format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     * @param a Alpha color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b, int a) {
        _textObject.setForeground(r, g, b, a);
    }

    /**
     * Setting text color of a ProgressBar in float RGB format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b) {
        _textObject.setForeground(r, g, b);
    }

    /**
     * Setting text color of a ProgressBar in float RGBA format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     * @param a Alpha color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b, float a) {
        _textObject.setForeground(r, g, b, a);
    }

    /**
     * Getting current text color.
     * 
     * @return Text color as as java.awt.Color.
     */
    public Color getForeground() {
        return _textObject.getForeground();
    }

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        return false;
    }

    /**
     * Adding item into the container (this).
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     */
    @Override
    public void addItem(IBaseItem item) {
        super.addItem(item);
        updateLayout();
    }

    /**
     * Setting item width. If the value is greater/less than the maximum/minimum
     * value of the width, then the width becomes equal to the maximum/minimum
     * value.
     * 
     * @param width Width of the item.
     */
    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        updateLayout();
    }

    /**
     * Setting X coordinate of the left-top corner of a shape.
     * 
     * @param x X position of the left-top corner.
     */
    @Override
    public void setX(int x) {
        super.setX(x);
        updateLayout();
    }

    void updateLayout() {
        updateProgressBar();
    }

    /**
     * Setting style of the ProgressBar.
     * <p>
     * Inner styles: "progressbar".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        setForeground(style.foreground);
        setFont(style.font);
        setTextAlignment(style.textAlignment);

        Style inner_style = style.getInnerStyle("progressbar");
        if (inner_style != null) {
            _rect.setStyle(inner_style);
        }
    }
}

File: SpaceVIL/Items/ProgressBar.cs
using System;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// LoadingScreen is designed to show progress 
    /// the execution of any long time task.
    /// <para/> Contains image and text.
    /// <para/> Supports all events except drag and drop.
    /// <para/> By default ProgressBar cannot get focus.
    /// </summary>
    public class ProgressBar : Prototype
    {
        static int count = 0;
        private TextLine _textObject;
        /// <summary>
        /// Setting the text that represents the progress of the unfinished task, 
        /// visible or invisible.
        /// </summary>
        /// <param name="value">True: if text should be visible. 
        /// False: if text should be invisible.</param>
        public void SetValueVisible(bool value)
        {
            _textObject.SetVisible(value);
        }
        /// <summary>
        /// Returns True if text that represents the progress of 
        /// the unfinished task is visible, otherwise returns False.
        /// </summary>
        /// <returns>True: if text is visible. 
        /// False: if text is invisible.</returns>
        public bool IsValueVisible()
        {
            return _textObject.IsVisible();
        }

        private Rectangle _rect;
        private int _maxValue = 100;
        private int _minValue = 0;
        private int _currentValue = 0;

        /// <summary>
        /// Default ProgressBar constructor.
        /// </summary>
        public ProgressBar()
        {
            SetItemName("ProgressBar_" + count);
            count++;

            _textObject = new TextLine();
            _textObject.SetItemName(GetItemName() + "_textObject");
            _rect = new Rectangle();

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.ProgressBar)));
            IsFocusable = false;
        }

        /// <summary>
        /// Initializing all elements in the ProgressBar. 
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            //text
            _textObject.SetItemText("0%");
            AddItems(_rect, _textObject);
        }

        /// <summary>
        /// Setting the maximum progress value of the unfinished task limit. 
        /// Progress value cannot be greater than this limit.
        /// </summary>
        /// <param name="value">Maximum progress value of the unfinished task limit.</param>
        public void SetMaxValue(int value)
        {
            _maxValue = value;
        }
        /// <summary>
        /// Getting the current maximum progress value of the unfinished task limit.
        /// </summary>
        /// <returns>Maximum progress value of the unfinished task limit.</returns>
        public int GetMaxValue()
        {
            return _maxValue;
        }
        /// <summary>
        /// Setting the minimum progress value of the unfinished task limit. 
        /// Progress value cannot be less than this limit.
        /// </summary>
        /// <param name="value">Minimum progress value of the unfinished task limit.</param>
        public void SetMinValue(int value)
        {
            _minValue = value;
        }
        /// <summary>
        /// Getting the current minimum sprogress value of the unfinished task limit.
        /// </summary>
        /// <returns>Minimum progress value of the unfinished task limit.</returns>
        public int GetMinValue()
        {
            return _minValue;
        }

        /// <summary>
        /// Setting the current progress value of the unfinished task. 
        /// If the value is greater/less than the maximum/minimum 
        /// progress value, then the progress value becomes equal 
        /// to the maximum/minimum value.
        /// </summary>
        /// <param name="value">Progress value of of the unfinished task.</param>
        public void SetCurrentValue(int value)
        {
            _currentValue = value;
            UpdateProgressBar();
        }
        /// <summary>
        /// Getting the progress value of the unfinished task.
        /// </summary>
        /// <returns>Progress value of the unfinished task.</returns>
        public int GetCurrentValue()
        {
            return _currentValue;
        }

        private void UpdateProgressBar()
        {
            float AllLength = _maxValue - _minValue;
            float DonePercent;
            _currentValue = (_currentValue > _maxValue) ? _maxValue : _currentValue;
            _currentValue = (_currentValue < _minValue) ? _minValue : _currentValue;
            DonePercent = (_currentValue - _minValue) / AllLength;
            string text = Math.Round(DonePercent * 100f, 1).ToString() + "%";
            _textObject.SetItemText(text);
            _rect.SetWidth((int)Math.Round(GetWidth() * DonePercent));
        }

        /// <summary>
        /// Setting alignment of the text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        /// <summary>
        /// Setting alignment of the text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as sequence of SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            _textObject.SetTextAlignment(alignment);
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to this ProgressBar.
        /// </summary>
        /// <param name="margin">Indents as SpaceVIL.Decorations.Indents.</param>
        public void SetTextMargin(Indents margin)
        {
            _textObject.SetMargin(margin);
        }
        /// <summary>
        /// Setting indents for the text to offset text relative to ProgressBar.
        /// </summary>
        /// <param name="left">Indent on the left.</param>
        /// <param name="top">Indent on the top.</param>
        /// <param name="right">Indent on the right.</param>
        /// <param name="bottom">Indent on the bottom.</param>
        public void SetTextMargin(int left = 0, int top = 0, int right = 0, int bottom = 0)
        {
            _textObject.SetMargin(left, top, right, bottom);
        }
        /// <summary>
        /// Getting indents of the text.
        /// </summary>
        /// <returns>Indents as SpaceVIL.Decorations.Indents.</returns>
        public Indents GetTextMargin()
        {
            return _textObject.GetMargin();
        }
        /// <summary>
        /// Setting font of the text.
        /// </summary>
        /// <param name="font">Font as System.Drawing.Font.</param>
        public void SetFont(Font font)
        {
            _textObject.SetFont(font);
        }
        /// <summary>
        /// Setting font size of the text.
        /// </summary>
        /// <param name="size">New size of the font.</param>
        public void SetFontSize(int size)
        {
            _textObject.SetFontSize(size);
        }
        /// <summary>
        /// Setting font style of the text.
        /// </summary>
        /// <param name="style">New font style as System.Drawing.FontStyle.</param>
        public void SetFontStyle(FontStyle style)
        {
            _textObject.SetFontStyle(style);
        }
        /// <summary>
        /// Setting new font family of the text.
        /// </summary>
        /// <param name="fontFamily">New font family as System.Drawing.FontFamily.</param>
        public void SetFontFamily(FontFamily fontFamily)
        {
            _textObject.SetFontFamily(fontFamily);
        }
        /// <summary>
        /// Getting the current font of the text.
        /// </summary>
        /// <returns>Font as System.Drawing.Font.</returns>
        public Font GetFont()
        {
            return _textObject.GetFont();
        }

        /// <summary>
        /// Getting the text width (useful when you need resize ProgressBar by text width).
        /// </summary>
        /// <returns>Text width.</returns>
        public int GetTextWidth()
        {
            return _textObject.GetWidth();
        }
        /// <summary>
        /// Getting the text height (useful when you need resize ProgressBar by text height).
        /// </summary>
        /// <returns>Text height.</returns>
        public int GetTextHeight()
        {
            return _textObject.GetHeight();
        }
        /// <summary>
        /// Setting text color of a ProgressBar.
        /// </summary>
        /// <param name="color">Text color as System.Drawing.Color.</param>
        public void SetForeground(Color color)
        {
            _textObject.SetForeground(color);
        }
        /// <summary>
        /// Setting text color of a ProgressBar in byte RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b)
        {
            _textObject.SetForeground(r, g, b);
        }
        /// <summary>
        /// Setting text color of a ProgressBar in byte RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        /// <param name="a">Alpha color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b, int a)
        {
            _textObject.SetForeground(r, g, b, a);
        }
        /// <summary>
        /// Setting text color of a ProgressBar in float RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b)
        {
            _textObject.SetForeground(r, g, b);
        }
        /// <summary>
        /// Setting text color of a ProgressBar in float RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        /// <param name="a">Alpha color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b, float a)
        {
            _textObject.SetForeground(r, g, b, a);
        }
        /// <summary>
        /// Getting current text color.
        /// </summary>
        /// <returns>Text color as System.Drawing.Color.</returns>
        public Color GetForeground()
        {
            return _textObject.GetForeground();
        }

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            return false;
        }

        /// <summary>
        /// Adding item into the container (this).
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public override void AddItem(IBaseItem item)
        {
            base.AddItem(item);
            UpdateLayout();
        }

        /// <summary>
        /// Setting item width. If the value is greater/less than the maximum/minimum 
        /// value of the width, then the width becomes equal to the maximum/minimum value.
        /// </summary>
        /// <param name="width"> Width of the item. </param>
        public override void SetWidth(int width)
        {
            base.SetWidth(width);
            UpdateLayout();
        }

        /// <summary>
        /// Setting X coordinate of the left-top corner of a shape.
        /// </summary>
        /// <param name="x">X position of the left-top corner.</param>
        public override void SetX(int x)
        {
            base.SetX(x);
            UpdateLayout();
        }

        internal void UpdateLayout()
        {
            UpdateProgressBar();
        }

        /// <summary>
        /// Setting style of the ContextMenu.
        /// <para/> Inner styles: "progressbar".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);

            SetForeground(style.Foreground);
            SetFont(style.Font);
            SetTextAlignment(style.TextAlignment);

            Style innerStyle = style.GetInnerStyle("progressbar");
            if (innerStyle != null)
            {
                _rect.SetStyle(innerStyle);
            }
        }
    }
}
